File: packages/typescript/src/verify.ts
/**
 * Signature Verification
 *
 * Verify ECDSA signatures using secp256k1 curve.
 */

import { sha256 } from 'js-sha256';
import { sha512 } from 'js-sha512';
import { ec as EC } from 'elliptic';
import { Signed, SignatureProof, VerificationResult } from './types';
import { toBytes } from './binary';

// Initialize secp256k1 curve
const ec = new EC('secp256k1');

/**
 * Verify a signed object
 *
 * @param signed - Signed object with value and proofs
 * @param isDataUpdate - Whether the value was signed as a DataUpdate
 * @returns VerificationResult with valid/invalid proof lists
 *
 * @example
 * ```typescript
 * const result = await verify(signedObject);
 * if (result.isValid) {
 *   console.log('All signatures valid');
 * }
 * ```
 */
export async function verify<T>(
  signed: Signed<T>,
  isDataUpdate: boolean = false
): Promise<VerificationResult> {
  // Compute the hash that should have been signed
  const bytes = toBytes(signed.value, isDataUpdate);
  const hashHex = sha256.hex(bytes);

  const validProofs: SignatureProof[] = [];
  const invalidProofs: SignatureProof[] = [];

  for (const proof of signed.proofs) {
    try {
      const isValid = await verifyHash(hashHex, proof.signature, proof.id);
      if (isValid) {
        validProofs.push(proof);
      } else {
        invalidProofs.push(proof);
      }
    } catch {
      // Verification error = invalid
      invalidProofs.push(proof);
    }
  }

  return {
    isValid: invalidProofs.length === 0 && validProofs.length > 0,
    validProofs,
    invalidProofs,
  };
}

/**
 * Verify a signature against a SHA-256 hash
 *
 * Protocol:
 * 1. Treat hash hex as UTF-8 bytes (NOT hex decode)
 * 2. SHA-512 hash
 * 3. Truncate to 32 bytes
 * 4. Verify ECDSA signature
 *
 * @param hashHex - SHA-256 hash as 64-character hex string
 * @param signature - DER-encoded signature in hex format
 * @param publicKeyId - Public key in hex (with or without 04 prefix)
 * @returns true if signature is valid
 */
export async function verifyHash(
  hashHex: string,
  signature: string,
  publicKeyId: string
): Promise<boolean> {
  try {
    // Step 1-2: Hash hex as UTF-8, then SHA-512
    const hexAsUtf8 = new TextEncoder().encode(hashHex);
    const sha512Hash = sha512.array(hexAsUtf8);

    // Step 3: Truncate to 32 bytes
    const truncatedHash = new Uint8Array(sha512Hash.slice(0, 32));

    // Normalize public key (add 04 prefix if needed)
    const fullPublicKey = normalizePublicKey(publicKeyId);

    // Step 4: Verify with elliptic
    const key = ec.keyFromPublic(fullPublicKey, 'hex');
    return key.verify(truncatedHash, signature);
  } catch {
    return false;
  }
}

/**
 * Verify a single signature proof against data
 *
 * @param data - The original data that was signed
 * @param proof - The signature proof to verify
 * @param isDataUpdate - Whether data was signed as DataUpdate
 * @returns true if signature is valid
 */
export async function verifySignature<T>(
  data: T,
  proof: SignatureProof,
  isDataUpdate: boolean = false
): Promise<boolean> {
  const bytes = toBytes(data, isDataUpdate);
  const hashHex = sha256.hex(bytes);
  return verifyHash(hashHex, proof.signature, proof.id);
}

/**
 * Normalize public key to full format (with 04 prefix)
 */
function normalizePublicKey(publicKey: string): string {
  // If 128 chars (without 04 prefix), add prefix
  if (publicKey.length === 128) {
    return '04' + publicKey;
  }
  // If 130 chars (with 04 prefix), return as-is
  if (publicKey.length === 130 && publicKey.startsWith('04')) {
    return publicKey;
  }
  // Otherwise return as-is
  return publicKey;
}

File: packages/python/src/constellation_sdk/verify.py
"""
Signature Verification.

Verify ECDSA signatures using secp256k1 curve.
"""

import hashlib
from typing import Any, TypeVar

from ecdsa import SECP256k1, VerifyingKey
from ecdsa.util import sigdecode_der

from .binary import to_bytes
from .hash import hash_bytes
from .types import SignatureProof, Signed, VerificationResult

T = TypeVar("T")


def verify(signed: Signed[T], is_data_update: bool = False) -> VerificationResult:
    """
    Verify a signed object.

    Args:
        signed: Signed object with value and proofs
        is_data_update: Whether the value was signed as a DataUpdate

    Returns:
        VerificationResult with valid/invalid proof lists

    Example:
        >>> result = verify(signed_object)
        >>> if result.is_valid:
        ...     print("All signatures valid")
    """
    # Compute the hash that should have been signed
    data_bytes = to_bytes(signed.value, is_data_update)
    hash_result = hash_bytes(data_bytes)

    valid_proofs: list[SignatureProof] = []
    invalid_proofs: list[SignatureProof] = []

    for proof in signed.proofs:
        try:
            is_valid = verify_hash(hash_result.value, proof.signature, proof.id)
            if is_valid:
                valid_proofs.append(proof)
            else:
                invalid_proofs.append(proof)
        except Exception:
            # Verification error = invalid
            invalid_proofs.append(proof)

    return VerificationResult(
        is_valid=len(invalid_proofs) == 0 and len(valid_proofs) > 0,
        valid_proofs=valid_proofs,
        invalid_proofs=invalid_proofs,
    )


def verify_hash(hash_hex: str, signature: str, public_key_id: str) -> bool:
    """
    Verify a signature against a SHA-256 hash.

    Protocol:
    1. Treat hash hex as UTF-8 bytes (NOT hex decode)
    2. SHA-512 hash
    3. Truncate to 32 bytes
    4. Verify ECDSA signature

    Args:
        hash_hex: SHA-256 hash as 64-character hex string
        signature: DER-encoded signature in hex format
        public_key_id: Public key in hex (with or without 04 prefix)

    Returns:
        True if signature is valid
    """
    try:
        # Step 1-2: Hash hex as UTF-8, then SHA-512
        hex_as_utf8 = hash_hex.encode("utf-8")
        sha512_hash = hashlib.sha512(hex_as_utf8).digest()

        # Step 3: Truncate to 32 bytes
        truncated_hash = sha512_hash[:32]

        # Normalize public key (add 04 prefix if needed)
        full_public_key = _normalize_public_key(public_key_id)

        # Step 4: Verify with ecdsa
        vk = VerifyingKey.from_string(bytes.fromhex(full_public_key), curve=SECP256k1)
        result: bool = vk.verify_digest(
            bytes.fromhex(signature), truncated_hash, sigdecode=sigdecode_der
        )
        return result
    except Exception:
        return False


def verify_signature(data: Any, proof: SignatureProof, is_data_update: bool = False) -> bool:
    """
    Verify a single signature proof against data.

    Args:
        data: The original data that was signed
        proof: The signature proof to verify
        is_data_update: Whether data was signed as DataUpdate

    Returns:
        True if signature is valid
    """
    data_bytes = to_bytes(data, is_data_update)
    hash_result = hash_bytes(data_bytes)
    return verify_hash(hash_result.value, proof.signature, proof.id)


def _normalize_public_key(public_key: str) -> str:
    """Normalize public key to full format (without 04 prefix for ecdsa library)."""
    # ecdsa library expects uncompressed key without 04 prefix
    if public_key.startswith("04") and len(public_key) == 130:
        return public_key[2:]
    if len(public_key) == 128:
        return public_key
    return public_key
